### DISCUSSION TITLE
How to get the raw string of the body of a request in a serverless function?

### DISCUSSION THREAD

**Discussion by leocabeza** [vercel/vercel#4524] [Help] [answered]
I'm working with [twitter webhooks](https://developer.twitter.com/en/docs/accounts-and-users/subscribe-account-activity/guides/securing-webhooks), and one of the steps for validating that the request comes from twitter is to read the raw body of the body in the request.

I've already tried: `JSON.stringify(request.body)` but that is a mismatch and I'm not able to verify.

Here are some related issues, but they are more _express_ related:
* https://stackoverflow.com/questions/53295279/securing-twitter-webhook
* https://twittercommunity.com/t/validating-the-webhook-signature-header-in-node-js/102525
* https://www.semicolonworld.com/question/49690/sha256-webhook-signature-from-woocommerce-never-verifies

**Comment by leocabeza** [comment]
Fore more info, I'm trying to do this in a serverless function: https://github.com/twitterdev/autohook/blob/master/index.js#L162

**Reply by wwwflorencio** [reply]
Are you using some HTTP framework or lib?

**Reply by wwwflorencio** [reply]
An option is to use express with serverless too, I have a function to work with twitter webhooks using express deployed on vercel.

https://github.com/eptaccio/twitter-dm-reply

For deploy questions see the [now.json file](https://github.com/eptaccio/twitter-dm-reply/blob/master/now.json)

**Reply by leocabeza** [reply]
I'm not using any http framework, just a plain serverless function.

**Reply by wwwflorencio** [reply]
[Following the docs](https://vercel.com/docs/v2/serverless-functions/introduction#an-example-node.js-serverless-function), is just access the `body` property

```javascript
module.exports = (req, res) => {
  res.json({
    body: req.body,
    query: req.query,
    cookies: req.cookies
  })
}
```

@leocabeza can you show a sample request?

**Reply by leocabeza** [reply]
When I receive a POST request, I can't get the raw string, because *req.body* is already parsed to an object

**Comment by styfle** [comment, marked as answer]
The `req.body` is parsed automatically by the platform for convenience ([docs](https://vercel.com/docs/runtimes?query=helpers#official-runtimes/node-js/node-js-request-and-response-objects/request-body)).

You can read the body directly using the Node.js Stream API since `req` is the [IncomingMessage](https://nodejs.org/docs/latest-v12.x/api/http.html#http_class_http_incomingmessage) object.

```js
module.exports = (req, res) => {
  if (req.method === 'POST') {
    let body = '';

    req.on('data', data => {
      body += data;
    });

    req.on('end', () => {
      console.log(body); // do stuff here
    });
  }
}
```
